### ISSUE TITLE
[feature] warn() when failing to auto-alter UID/EID

### ISSUE THREAD

**Issue by FGasper** [Perl/perl5#18135]
Currently when I run this:
```
perl -e'$> = 1; { local $> = 99; }; print "euid: $>\n"'
```
… it prints `99`. The kernel forbids the process from setting EUID back to 1, so it’s not a Perl _bug_ that the EUID is still 99; however, it’s pretty confusing that there’s no notification of Perl’s internal failure to reset EUID back to 1.

Would it be considered useful if I submitted a patch to create such a warning?

**Comment by Grinnz** [comment]
As [documented](https://perldoc.pl/variables/$%3E), you must always check $! for errors after setting any of `$>`, `$<`, `$)`, or `$(`. I prefer using the equivalent functions like [POSIX::setuid](https://perldoc.pl/POSIX#setuid) which are easier to check for success (though unfortunately undocumented how to do so, they return undef on failure and the error is again set in `$!`).

**Comment by FGasper** [comment]
@Grinnz Both of the manual `$>` sets _are_ succeeding. The one that’s failing is the one that tries to set $> from 99 to 1, which there seems to be no reliable way to check.

**Comment by Grinnz** [comment]
Regardless of whether it's implicit or explicit the only way to determine if `$>` assignment succeeded is to check `$!` immediately after it occurs.

**Comment by FGasper** [comment]
Expecting Perl code after a block to do an explicit error-check of an implicit system call:
```
perl -e'$> = 1; { local $> = 99; }; print "euid: $> ($!)\n"'
```
… seems like a weird ask, the kind of thing more reasonably communicated via push than pull. <shrug>

**Comment by Grinnz** [comment]
I totally agree it's non-ideal, but like I said my solution is generally not to assign to these variables. I would probably use a scope guard that attempts to set it back on exit using POSIX and dies on error.

Regarding the actual feature request, I think it could be useful to warn on failures but on the other hand the 'exec' warnings category which is used for a similar purpose is more annoying than useful. So I don't really have an opinion either way.

**Comment by Leont** [comment]
Actually, this very much looks like a perl bug. What's happening is this:
```
setresuid(-1, 1, -1)                    = 0
setresuid(-1, 0, -1)                    = 0
setresuid(-1, 99, -1)                   = 0
setresuid(-1, 1, -1)                    = -1 EPERM (Operation not permitted)
```
That second setresuid (caused by the `local $>`) shouldn't be there. If it wasn't there the third `setresuid` would fail (as it should) instead of the fourth.

**Comment by FGasper** [comment]
@Leont Thank you … I was wondering why only 99->1 was failing!

**Comment by Leont** [comment]
> @Leont Thank you … I was wondering why only 99->1 was failing!

I know how to fix this, but I'm worried about breaking other things in the process. Especially because security is involved.

**Comment by FGasper** [comment]
I was wondering about that, too … if this is a “load-bearing bug”.

**Comment by xsawyerx** [comment]
@Leont, is there anything we can do to try and validate whether this breaks other things?

@lightsey, could you also provide your thoughts on the security perspective of this?

**Comment by Leont** [comment]
Currently `local $>` means the same as `local $> = 0` (as all localizations first undefine the value before any assignment happens). Fixing this would involve essentially special casing it to mean roughly `local $> = $>` (by skipping the set-magic).

If someone does `local $>` meaning to be `local $> = 0`, for example to temporarily restore root privileges, their code would break. At least it's the non-catastrophic direction of breaking (less permissions not more).

I don't see how it would break the other way around, but if it would that would be a serious problem.

**Comment by CowboyTim** [comment]
If this is a perl bug, the fact that `local $>` first unsets to 0, is what makes the first setresuid call actually succeed as the EUID is 0 temporary  - given that the script as all the permissions of course.

Fixing this at the end of the block by again going temporarily to 0 should be no issue then - compared to when someone does just a `local $>`.

    $ perl -e'$> = 1; { local $> = 99; $>=0 }; print "euid: $> ($!)\n"'
    euid: 1 ()

The case of

    $ perl -e'$> = 1; { local $> = $>; }; print "euid: $> ($!)\n"'
    euid: 0 ()

Is indeed even plain wrong and a severe security risk, as the process nicely continuous as root, not EUID=1 or EUID=99!

**Comment by Leont** [comment]
I hadn't even realized that second case yet. That is really unfortunate behavior :-/.
